docs/ROOTFS_DISCOVERY.md

```diff
@@ -0,0 +1,395 @@
+---
+title: Boot Components & Root File System Discovery
+category: Booting
+layout: default
+SPDX-License-Identifier: LGPL-2.1-or-later
+---
+
+# Boot Components & Root File System Discovery
+
+The recommended way to boot a [`systemd`](https://systemd.io/) based
+[UEFI](https://en.wikipedia.org/wiki/UEFI) system consists primarily of three
+components:
+
+1. A boot loader,
+   i.e. [`systemd-boot`](https://www.freedesktop.org/software/systemd/man/latest/systemd-boot.html)
+   that provides interactive and programmatic control of what precisely to
+   boot. It takes care of enumerating all possible boot targets, potentially
+   presenting it to the user in a menu, but otherwise picking an item
+   automatically, implementing boot counting and automatic rollback if desired.
+
+2. A unified kernel image ("UKI"), i.e. an UEFI PE executable that combines
+   [`systemd-stub`](https://www.freedesktop.org/software/systemd/man/latest/systemd-stub.html),
+   a Linux kernel, and an initial RAM disk ("`initrd`") into one. UKIs are
+   self-descriptive: the aforementioned boot loader enumerates these UKIs and
+   automatically extracts all information necessary to determine which menu
+   entries to generate for them. Within the UKI runtime the transition from the
+   UEFI firmware world into the Linux world takes place, i.e. it encapsulates
+   the fundamental `ExitBootServices()` UEFI call, that ends PC firmware
+   control, and let's the Linux kernel take over.
+
+3. A root file system ("`rootfs`"): this is where the regular OS is
+   located. The primary job of the early userspace that is contained in the
+   `initrd` that itself is part of the UKI, is to find, set up and pivot into
+   the `rootfs`.
+
+> [!NOTE]
+> The above is how `systemd` upstream recommends a system is put together. However,
+> distributions differ from this, sometimes massively – in particular when
+> their focus is on supporting legacy (i.e. non-UEFI) hardware platforms. We
```

Comment:
I wouldn't equate non-UEFI to legacy platforms and think phrasing it the other way around would be closer to reality "...supporting non-UEFI (i.e. legacy, and embedded) hardware platforms.".

Yes, many distros use uboot to shoehorn UEFI into embedded platforms but not doing so does not mean that they are legacy platforms.

Reply:
non-UEFI are definitely legacy, modern embedded like arm64, riscv and loong64 can do uefi just fine

Reply:
Server and consumer hardware sure, but industrial SoCs no because they only provide a basic stage 0 bootloader and would need to rely on uboot et al. to provide a UEFI environment

Reply:
That's their problem - they can still provide uefi for new platforms if they want to, and if they don't, they are legacy